packages/big-design/src/components/Tree/Tree.tsx

```diff
@@ -1,53 +1,56 @@
-import React, { useMemo, useReducer, useState } from 'react';
+import React, { createContext, useEffect, useMemo, useState } from 'react';
 
-import { useUniqueId } from '../../hooks';
-import { typedMemo } from '../../utils';
-
-import { createReducer, createReducerInit } from './reducer';
 import { StyledUl } from './styled';
 import { TreeNode } from './TreeNode';
-import { TreeProps } from './types';
+import { TreeContextState, TreeProps } from './types';
+
+// We don't have access to the type yet, so we need to pass any into the generic.
+export const TreeContext = createContext<TreeContextState<any>>({
+  expandable: {
+    expandedNodes: [],
+  },
+  focusable: {
+    focusedNode: '',
+    onFocus: () => null,
+  },
+  onKeyDown: () => null,
+});
 
-const InternalTree = <T extends unknown>({
+export const Tree = <T extends unknown>({
+  disabledNodes,
+  expandable,
+  focusable,
   iconless,
   id,
-  initialNodes,
-  onCollapse,
-  onExpand,
-  onSelect,
+  nodes,
+  onKeyDown,
+  onNodeClick,
   selectable,
 }: TreeProps<T>): React.ReactElement<TreeProps<T>> => {
-  const [nodes] = useState(initialNodes);
-  const uniqueTreeId = useUniqueId('tree');
-  const treeId = id ?? uniqueTreeId;
-  const reducer = useMemo(() => createReducer<T>(), []);
-  const reducerInit = useMemo(() => createReducerInit<T>(), []);
-  const [state, dispatch] = useReducer(reducer, { nodes, selectable }, reducerInit);
+  const [internalNodes, setInternalNodes] = useState(nodes);
+  const initialTreeContext: TreeContextState<T> = {
+    disabledNodes,
+    expandable,
+    focusable,
+    iconless,
+    onKeyDown,
+    onNodeClick,
+    selectable,
+  };
 
-  const renderedItems = useMemo(
-    () =>
-      nodes.map((node, index) => (
-        <TreeNode
-          {...node}
-          dispatch={dispatch}
-          iconless={iconless}
-          key={index}
-          onCollapse={onCollapse}
-          onExpand={onExpand}
-          onSelect={onSelect}
-          selectable={selectable}
-          state={state}
-          treeId={treeId}
-        />
-      )),
-    [iconless, nodes, onCollapse, onExpand, onSelect, selectable, state, treeId],
-  );
+  useEffect(() => {
```

Comment:
I see you are using this pattern in multiple places. Consider abstracting it to a `useDerivedState` hook.

You can also optimize it a bit, `useEffect` runs after browser paint, you can use another `useState` to keep track of the previous props value. This will trigger a re-render before commiting the changes to the DOM. [More info](https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops).

Reply:
We tried this out and were running into infinite rerenders. Just noting we can probably come back to this if performance is a huge issue with this component.